fastapi/routing.py

```diff
@@ -123,6 +123,11 @@ async def serialize_response(
             exclude_defaults=exclude_defaults,
             exclude_none=exclude_none,
         )
+        try:
```

Comment:
response_content is not necessarily a dictionary. If you remove the try catch some tests fail.

Reply:
Then why don't you check the variable data type instead?

Reply:
For two reasons:
1. Python advocates EAFP
2. For all I know it may not be a dict, it could be a dict like object. 

I didn't follow the code far enough to see what all could get passed, but we're dealing with Pydantic models upstream and it's perfectly possible to implement this dict functionality. 

Though to be consistent I note im breaking EAFP inside the try catch, so I could be 
```
try:
    response_content = response_content["__root__"]
except TypeErrror, KeyError:
    pass
```

But I think the intention is slightly less clear.

Reply:
Actually, there's already an `if isinstance(res, BaseModel)` check in `_prepare_response_content` that might be a better place for this code, because presumably it should apply if (and only if) it's dealing with a Pydantic `BaseModel` instance.

Somewhat relatedly, it looks like Pydantic itself does [something similar in the `BaseModel.json()` method](https://github.com/patrickkwang/pydantic/blob/bb22ec78b318dea5e7f85f73e807b783b99cdc2b/pydantic/main.py#L400-L433), where it notices that the model has a custom `__root__` and then extracts the nested value:
```python
    def json(
        self,
        ...
    ) -> str:
        ... # (trimmed for brevity)

        data = self.dict(
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            exclude_unset=exclude_unset,
            exclude_defaults=exclude_defaults,
            exclude_none=exclude_none,
        )
        if self.__custom_root_type__:
            data = data[ROOT_KEY]
```
.. where `ROOT_KEY == "__root__"`, and `self.__custom_root_type__` is a property of all `pydantic.BaseModel`s (possibly None, but defined either way).

So putting all of this together - combining the logic from `BaseModel.json` with the approach in this PR, applied to @cikay's feedback:
```diff
 def _prepare_response_content(
     res: Any,
     *,
     exclude_unset: bool,
     exclude_defaults: bool = False,
     exclude_none: bool = False,
 ) -> Any:
     if isinstance(res, BaseModel):
         read_with_orm_mode = getattr(res.__config__, "read_with_orm_mode", None)
         if read_with_orm_mode:
             # Let from_orm extract the data from this model instead of converting
             # it now to a dict.
             # Otherwise there's no way to extract lazy data that requires attribute
             # access instead of dict iteration, e.g. lazy relationships.
             return res
-        return res.dict(
+        res_dict = res.dict(
             by_alias=True,
             exclude_unset=exclude_unset,
             exclude_defaults=exclude_defaults,
             exclude_none=exclude_none,
         )
+        if res.__custom_root_type__:
+            res_dict = res_dict["__root__"]
+        return res_dict
```

... or something like that (I haven't been able to try running this yet).

The fact that `pydantic.BaseModel` does basically the same exact thing might be a sign that it's a valid, non-crazy approach here to have well-defined behavior around types with custom `__root__`.